net: socket: syscall for socketpair(2)

cfriedt opened the pull request:
Provide an implementation for socketpair(2). Currently, this is limited to the AF_UNIX address family, SOCK_STREAM, SOCK_RAW, and protocol 0.

Please feel free to suggest additional testing in `tests/net/socket/socketpair/` or to suggest improvements in the implementation.

Fixes #24366

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+
+	/* Zero-initializing this (with the exception of setting the family)
+	 * has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
+         *
```
Some indentation issue on this line

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+	temp_sv[0] = res;
+
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
```
Please write this `if (res == -1) {` in order to follow the same style as the other functions in this file.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+	res = z_impl_zsock_socket(family, type, proto);
+	if (-1 == res) {
+		return -1;
+	}
```
Please add empty line after }

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+		return -1;
+	}
+
+	sv[ 0 ] = temp_sv[ 0 ];
```
no spaces inside `[]` so writing this as `sv[0] = temp_sv[0]` is ok

jukkar commented on tests/net/socket/socketpair/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
Please put the tests to separate commit

jukkar commented on tests/net/socket/socketpair/src/main.c:
```diff
@@ -0,0 +1,76 @@
+
+#include <net/socket.h>
+
+//#include "../../socket_helpers.h"
```
Please remove this commented out line

jukkar commented on tests/net/socket/socketpair/src/main.c:
```diff
@@ -0,0 +1,76 @@
+
+//#include "../../socket_helpers.h"
+
+#ifndef ARRAY_SIZE
```
just include sys/util.h which will provide ARRAY_SIZE

jukkar commented on tests/net/socket/socketpair/src/main.c:
```diff
@@ -0,0 +1,76 @@
+
+	ztest_test_suite(socket_socketpair,
+			 ztest_user_unit_test(test_socket_socketpair_open_close));
+
```
We need to add a test that sends/receives data and makes sure that the data is passed properly.

jukkar reviewed (changes requested):
Looks good, minor nits. We also need to add more tests.

pfalcon commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+	int temp_sv[2];
+
+	/* Zero-initializing this (with the exception of setting the family)
+	 * has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
```
> Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.

Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this is required, or let me conjecture that this is serious security issue, where you open your "inter-thread communication" (per https://github.com/zephyrproject-rtos/zephyr/issues/24366) to attacks from the wide Internet.

Actually, you unlikely to provide such a reference, because normative spec says essebtially the opposite:

> The socketpair() function shall create an **unbound** pair of connected sockets

(emphasis mine). So, let me give -1 to this for now, and let's think how to deal with that.

cfriedt replied:
Hmm... Interesting. Normally when I look up this man page, I see that it says "unnamed" rather than "unbound". E.g. in Linux, it only works with AF_UNIX.

Also, I just realized I'm missing a listen(2) call, which would normally be necessary for e.g. AF_INET / SOCK_STREAM so it likely would not work in those cases and it would also expose a race condition / security issue, unless there was a way, as you say, to connect two sockets without binding.

cfriedt replied:
I would actually prefer to only support AF_UNIX, but since there is no such support in Zephyr yet, I thought I should work with what is available.

pfalcon replied:
> I would actually prefer to only support AF_UNIX

That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:

1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.

Feel free to give this idea a thought.

cfriedt replied:
@pfalcon - if you would prefer as well, I could remove AF_INET and focus on adding AF_UNIX support instead.

jukkar replied:
> I could remove AF_INET and focus on adding AF_UNIX support instead. Its use could be limited to socketpair() at least initially

+1 to this

cfriedt replied:
> > I would actually prefer to only support AF_UNIX
> 
> That would be natural course of thinking, and then socketpair() implementation wouldn't be part of network subsystem per se.

true...

> As you know, native Zephyr way to communicate data streams between threads are Zephyr pipes. If I were to implement this in the most useful way, and yet bound the scope-creep, I'd consider following way:
> 
> 1. See if socketpair(AF_UNIX) can be reasonably implemented in terms of classical pipe().
> 2. If so, implement pipe() for POSIX subsys (I don't think we have it), and then socketpair() in top of it.
> 
> Feel free to give this idea a thought.

Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr. Are there additional / special markers that need to be inserted in order to support SOCK_DGRAM?

pfalcon replied:
> Good idea. I wasn't actually sure what would be the best way to implement it in Zephyr.

@cfriedt, before discussing this further, I'd like to get your reply to https://github.com/zephyrproject-rtos/zephyr/issues/24366#issuecomment-613965163 . Because again, if you want to just signal an event to poll(), then eventfd() is a better choice.

(OTOH, if you really want to support SOCK_DGRAM (in addition to natural choice of SOCK_STREAM), then might be then reusing existing network socket infrastructure actually makes sense after all (or it would be indeed a bunch of code duplication between network socket subsystem and non-network socket impl).)

cfriedt replied:
@pfalcon - I would be fine just supporting SOCK_STREAM (SOCK_RAW is even probably unnecessary and I don't even know if there is a difference to SOCK_STREAM in this case, but some people prefer it for AF_UNIX).

cfriedt replied:
> > Zero-initializing this (with the exception of setting the family) has the same effect as using INADDR_ANY_INIT or IN6ADDR_ANY_INIT.
> 
> Perhaps. But why do you bind to INADDR_ANY? Please provide normative reference that this 

Originally, I was just thinking INADDR_LOOPBACK, but then saw that `lo` requires an additional CONFIG option, but yea.. let's just forget about AF_INET.

pfalcon commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,76 @@ int z_impl_zsock_socket(int family, int type, int proto)
+
+	res = z_impl_zsock_bind(temp_sv[1], sa, len);
+	if (-1 == res) {
+		z_impl_zsock_close(temp_sv[0]);
```
IMHO, repetitive close() call should be replaced with "C exception hanling with goto" pattern.

cfriedt replied:
I prefer that too. Some people are really against goto though.
 Will rework.

pfalcon reviewed (changes requested):
There's an obvious security issue with this patch, see inline comment.

jukkar commented on include/net/net_ip.h:
```diff
@@ -208,6 +213,12 @@ struct sockaddr_ll_ptr {
+/** Socket address for UNIX domain sockets, pipes, etc */
+struct sockaddr_un {
+	sa_family_t sun_family;    /* AF_UNIX  */
+	char        sun_path[108]; /* Pathname */
```
I think we need to make this configurable via Kconfig option. The value 108 is the same as in Linux but for Zephyr it is consuming way too much memory. For socketpair use, the path is not used according to manual `Likewise, the two sockets created by socketpair(2) are unnamed.  When the address of an unnamed socket is returned, its length is sizeof(sa_family_t), and sun_path should not be inspected.`
We could make the default value for path length so that sockaddr_un and sockaddr_in6 would be the same size. This way there is some space for path and user can increase the value if needed.

cfriedt replied:
@jukkar - yeah, I agree, it is a bit much, and technically speaking, it is unnecessary for socketpair(2) or pipe(2). Would it be better to remove this definition, since named sockets are not even part of the proposal? Originally, I was going to use the sun_path as either an actual path or a string of random bytes to assist in matching fds, but both of those things are probably unnecessary.

ceolin commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,46 @@ int z_impl_zsock_socket(int family, int type, int proto)
+	return res;
+}
+
 #ifdef CONFIG_USERSPACE
```
I think it is missing the verification handler for this syscall. If that's correct your test is probably no calling this syscall from usermode

cfriedt replied:
Good catch! I think I have it now.

ceolin commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -173,6 +175,46 @@ int z_impl_zsock_socket(int family, int type, int proto)
+	return res;
+}
+
 #ifdef CONFIG_USERSPACE
```
I think it is missing the verification handler for this syscall. If that's correct your test is probably no calling this syscall from usermode

ceolin commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -184,6 +226,17 @@ static inline int z_vrfy_zsock_socket(int family, int type, int proto)
+{
+	/* implementation call to net_context_get() should do all necessary
+	 * checking
+	 */
```
You have to check if the calling thread has permission to `sv`.  So, here you have to create a temporary copy, call the implementation with this copy and after it return you copy the memory back using `z_user_to_copy`

ceolin replied:
```
int ret;
int tmp[2];

ret = z_impl_zsock_socketpair(family, type, proto, tmp);
Z_OOPS(z_user_to_copy(sv, tmp, sizeof(tmp));
return ret;
```

cfriedt replied:
Thanks @ceolin !

cfriedt commented:
If it's ok with you all (@pfalcon, @jukkar, @ceolin, @tbursztyka), I will leave this PR open and implement pipe(2) separately. I believe this review will be fairly cut & dry once pipe(2) is implemented.

Also, do you know of a simple way I can run my tests natively or in Qemu? So far all I have been contributing to is BLE.

Sorry for all of the whitespace issues - clang-format does not do the trick (#23823) but uncrustify might (#21392)

cfriedt commented:
It's probably better to not use pipe(2) for socketpair(2) since a pair of sockets form a bidirectional channel rather than a unidirectional channel. Likely, will use a pair of k_fifo. Technically, it is possible using pipe(2) but that would mean 4 file descriptors would need to be created instead of only 2.

cfriedt committed:
net: socket: syscall for socketpair(2)

Provide an implementation for socketpair(2). Currently, this is limited to the
AF_UNIX address family, SOCK_STREAM, SOCK_RAW, and protocol 0.

Fixes #24366

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
stashing

cfriedt committed:
fixed prj.conf

cfriedt committed:
still encountering spinlock error

cfriedt committed:
works now with the following patch by pabigot

diff --git a/kernel/pipes.c b/kernel/pipes.c
index efd2855c5e..c6ae71555d 100644
--- a/kernel/pipes.c
+++ b/kernel/pipes.c
@@ -135,10 +135,11 @@ void k_pipe_init(struct k_pipe *pipe, unsigned char *buffer, size_t size)
        pipe->bytes_used = 0;
        pipe->read_index = 0;
        pipe->write_index = 0;
-       pipe->flags = 0;
+       pipe->lock = (struct k_spinlock){};
        z_waitq_init(&pipe->wait_q.writers);
        z_waitq_init(&pipe->wait_q.readers);
        SYS_TRACING_OBJ_INIT(k_pipe, pipe);
+       pipe->flags = 0;
        z_object_init(pipe);
 }

cfriedt committed:
write(2), read(2), send(2), recv(2), sendto(2), recvfrom(2) working

Except write / read need to be macro'ed out to send / recv due to Zephyr's
POSIX naming, otherwise they attempt to use write / read from stdio.

Currently only working in the sv[0] -> sv[1] direction.

If I attempt to write in the sv[1] -> sv[0] direction, I encounter the
same error that I encountered before when there was an assertion verifying
the pipe's spinlock.

Output:

ninja: Entering directory `build'
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: cortex-m3
qemu-system-arm: warning: nic stellaris_enet.0 has no peer
*** Booting Zephyr OS build zephyr-v2.2.0-1661-g0cc6bdf36c5c  ***
Running test suite socket_socketpair
===================================================================
starting test - test_socket_socketpair_happy_path
calling socketpair(AF_LOCAL, SOCK_STREAM, 0, sv)
data direction: 0 -> 1
testing write(2)
testing read(2)
testing send(2)
testing recv(2)
testing sendto(2)
testing recvfrom(2)
data direction: 1 -> 0
testing write(2)
ASSERTION FAIL [z_spin_lock_valid(l)] @ ../include/spinlock.h:44
	Recursive spinlock 0x20000bf4
E: r0/a1:  0x00000004  r1/a2:  0x0000002c  r2/a3:  0x0000000a
E: r3/a4:  0x00000000 r12/ip:  0x01010101 r14/lr:  0x00017c39
E:  xpsr:  0x41000000
E: Faulting instruction address (r15/pc): 0x0001d002
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x200009c0 (unknown)
E: Halting system
qemu-system-arm: terminating on signal 2
ninja: build stopped: interrupted by user.

Stack Trace:

 #0  arch_irq_lock () at ../include/arch/arm/aarch32/asm_inline_gcc.h:56
 #1  k_spin_lock (l=0x20000bf4 <kheap_poolheap.heap_mem_pool+104>) at ../include/spinlock.h:41
 #2  z_pipe_put_internal (pipe=0x20000be0 <kheap_poolheap.heap_mem_pool+84>, async_desc=0x0 <crc32_ieee>, data=0x2879c "Hello, socketpair(2) world!", bytes_to_write=27, bytes_written=0x20001e90 <ztest_thread_stack+648>, min_xfer=0, timeout=0) at /home/cfriedt/workspace/zephyrproject/zephyr/kernel/pipes.c:449
 #3  0x000261da in z_impl_k_pipe_put (pipe=0x20000be0 <kheap_poolheap.heap_mem_pool+84>, data=0x2879c, bytes_to_write=27, bytes_written=0x20001e90 <ztest_thread_stack+648>, min_xfer=0, timeout=0) at /home/cfriedt/workspace/zephyrproject/zephyr/kernel/pipes.c:741
 #4  0x0001d24e in k_pipe_put (pipe=0x20000be0 <kheap_poolheap.heap_mem_pool+84>, data=0x2879c, bytes_to_write=27, bytes_written=0x20001e90 <ztest_thread_stack+648>, min_xfer=0, timeout=0) at zephyr/include/generated/syscalls/kernel.h:908
 #5  0x0000808a in spair_write (obj=0x20000bec <kheap_poolheap.heap_mem_pool+96>, buffer=0x2879c, count=27) at /home/cfriedt/workspace/zephyrproject/zephyr/subsys/net/lib/sockets/socketpair.c:217
 #6  0x0001d442 in spair_sendto (obj=0x20000bec <kheap_poolheap.heap_mem_pool+96>, buf=0x2879c, len=27, flags=0, dest_addr=0x0 <crc32_ieee>, addrlen=0) at /home/cfriedt/workspace/zephyrproject/zephyr/subsys/net/lib/sockets/socketpair.c:294
 #7  0x0001d1d2 in z_impl_zsock_sendto (sock=1, buf=0x2879c, len=27, flags=0, dest_addr=0x0 <crc32_ieee>, addrlen=0) at /home/cfriedt/workspace/zephyrproject/zephyr/subsys/net/lib/sockets/sockets.c:566
 #8  0x0001c1f0 in zsock_sendto (sock=1, buf=0x2879c, len=27, flags=0, dest_addr=0x0 <crc32_ieee>, addrlen=0) at zephyr/include/generated/syscalls/socket.h:137
 #9  0x0001bf88 in zsock_send (sock=1, buf=0x2879c, len=27, flags=0) at /home/cfriedt/workspace/zephyrproject/zephyr/include/net/socket.h:293
 #10 0x00006c18 in happy_path_inner (family=6, family_s=0x28ee8 "AF_LOCAL", type=1, type_s=0x28efc "SOCK_STREAM", proto=0, proto_s=0x28f14 "0") at /home/cfriedt/workspace/zephyrproject/zephyr/tests/net/socket/socketpair/src/main.c:58
 #11 0x0000700c in test_socket_socketpair_happy_path () at /home/cfriedt/workspace/zephyrproject/zephyr/tests/net/socket/socketpair/src/main.c:185

pfalcon commented:
@cfriedt: Same comment as for the other PR - please deal with checkstyle issues.

(Also, to set expectations right - there was a talk about release readiness meeting. Which reminds that we may be approaching a pre-release freeze. Whatever isn't merged before it, may need to wait a month or so until next devel window is open. Surely feel free to work at you pace. What I wanted to say is that if you don't get a prompt review (once CI passes), don't hesitate to ping me. Thanks.)

cfriedt committed:
stashing

cfriedt commented:
Going to close the PR for the next while. Will reopen soon.
